### PR TITLE
Pydantic gene

### PR DESCRIPTION
**Pull request by northwestwitch**
This PR adds a functionality or fixes a bug.
- Converted the HgncGene class into a pydantic class

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

**How to test**:
1. how to test it, possibly with real cases/data

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [x] code approved by DN
- [x] tests executed by CR

### CODE REVIEW

**Comment by northwestwitch** [comment]
Tested with some scout update files  that I had locally (downloaded May 28th):

##" Loading genes with main branch:

image


##" Loading genes with main branch:

image

Comparing POT1 gene documents saved in db by both branches:

image

File: scout/build/genes/hgnc_gene.py [hunk 1]
```diff
@@ -16,137 +15,9 @@ def build_phenotype(phenotype_info):
...
-        gene_obj["phenotypes"] = phenotype_objs
-
-    for key in list(gene_obj):
-        if gene_obj[key] is None:
-            gene_obj.pop(key)
-
-    return gene_obj
+    gene_info["build"] = build
+    hgnc_gene = HgncGene(**gene_info)
+    return hgnc_gene.model_dump(exclude_none=True)
```
[the remark this answers is not in the document]
Comment (dnil) [reply scout/build/genes/hgnc_gene.py:23]: Good stuff! Did you check ds uses for existing values by any chance? Otw we will find out! 😬

File: tests/adapter/conftest.py [hunk 2]
```diff
@@ -59,7 +59,7 @@ def parsed_gene():
         "hgnc_id": 1,
         "hgnc_symbol": "AAA",
         "ensembl_id": "ENSG1",
-        "chrom": "1",
+        "chromosome": "1",
```
Comment (dnil) [inline tests/adapter/conftest.py:62]: Hm, yes, I think this is correct - the `chrom` should all have been temporary shorthands. I will say it is a bit confusing to have `chrom` for `Exon`, `HgncTranscript` etc, but we are in a small majority of "chromosome" for storage so go.

File: tests/build/test_build_hgnc_gene.py [hunk 3]
```diff
@@ -50,17 +50,16 @@ def test_build_hgnc_gene_inappropriate_type(test_gene, key):
     # WHEN setting key to None
     test_gene[key] = None
     # THEN calling build_hgnc_gene() will raise TypeError
-    with pytest.raises(TypeError):
+    with pytest.raises(ValidationError):
```
Comment (dnil) [inline tests/build/test_build_hgnc_gene.py:53]: 👍

File: tests/conftest.py [hunk 4]
```diff
@@ -125,16 +125,18 @@ def mock_redirect():
...
+            ensembl_id="ENSG00000176022",
+            chromosome="1",
+            start=1232237,
+            end=1235041,
+            build="38",
+        )
     )
 
-    return gene
+    return gene.model_dump()
```
Comment (dnil) [inline tests/conftest.py:139]: 👍

**Review by dnil — COMMENTED** [review]
✅

File: tests/adapter/conftest.py (hunk 2 above)
Comment (northwestwitch) [reply tests/adapter/conftest.py:62]: It gets saved as chromosome in the database, it was just the passages before, it's first chromosome, then [becomes chrom](https://github.com/Clinical-Genomics/scout/blob/415bfff6d84eb76765ba5599065100d17029843f/scout/build/genes/hgnc_gene.py#L100), to be [converted again](https://github.com/Clinical-Genomics/scout/blob/415bfff6d84eb76765ba5599065100d17029843f/scout/models/hgnc_map.py#L126) in chromosome in the model. Not confusing at all!

File: scout/build/genes/hgnc_gene.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/build/genes/hgnc_gene.py:23]: Was this your question?

File: tests/adapter/conftest.py (hunk 2 above)
Comment (dnil) [reply tests/adapter/conftest.py:62]: Exactly, keeps you on your toes, doesn’t it! 😆 There may be something to different names names for some things in disk/db/mem by need or to quickly tell
things that are further decorated in memory apart. But different names to member vars that store the same is not it.

File: scout/build/genes/hgnc_gene.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dnil) [reply scout/build/genes/hgnc_gene.py:23]: Yes, we’ll, you remove a slightly larger subset now compared to before, so my question was if you followed some of the now possibly non existing keys to see what happens…

[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/build/genes/hgnc_gene.py:23]: Ah no! :)

[the remark this answers is not in the document]
Comment (dnil) [reply scout/build/genes/hgnc_gene.py:23]: Might be safest to at least randomly pick a couple. But, hey, the tests pass, and the db loads, so how bad can it be. 😅 I will sure borrow that `exclude_none` - much cleaner than selectively removing certain keys afterwards.

**Comment by northwestwitch** [comment]
I'll try loading the genes on hasta and do some browsing on the portal

File: tests/conftest.py (hunk 4 above)
Comment (dnil) [reply tests/conftest.py:139]: You could try the `exclude_none` a bit further here if you like. Not going to go the full distance, but...

Comment (northwestwitch) [reply tests/conftest.py]: Done

**Comment by northwestwitch** [comment]
Loading on Hasta: 🆗 

image

**Comment by northwestwitch** [comment]
Browsing some genes and using them for filtering of creating HPO panels etc it seems that nothing changes in comparison with prod

**Review by dnil — APPROVED** [review]
🆗

**Pull request merged**